File: api/models/email.py
from utils import parse_mongo
from bson.objectid import ObjectId

category_names = ['starred', 'spam', 'inbox', 'important', 'sent', 'trash']

def create(db, subject, content, from_user, to_user):
    email = { 'subject': subject, 'content': content, 'from_user': ObjectId(from_user), 'is_draft': True }
    if to_user:
        email = { 'subject': subject, 'content': content, 'from_user': ObjectId(from_user), 'to_user': ObjectId(to_user), 'is_draft': True }
    return db.emails.insert_one(email)

def updateEmail(db, user_id, subject, email_id, content, to_user, is_draft):
    db.emails.update_one({'_id': ObjectId(email_id)}, { '$set': {'subject': subject, 'content': content, 'is_draft': is_draft }})
    if to_user:
        db.emails.update_one({'_id': ObjectId(email_id)}, { '$set': {'subject': subject, 'content': content, 'is_draft': is_draft, 'to_user': ObjectId(to_user) }})
    if is_draft:
        db.users.update_one({ '_id': ObjectId(user_id) }, { '$addToSet': { 'draft': ObjectId(email_id) } })

def removeFromAllCategories(db, user_id, email_id):
    for cat in category_names:
        db.users.update_one({ '_id': ObjectId(user_id) }, { '$pull': { cat: ObjectId(email_id) } })

def addToCategory(db, user_id, email_id, category):
    if category not in category_names:
        return
    removeFromAllCategories(db, user_id, email_id)
    db.users.update_one({ '_id': ObjectId(user_id) }, { '$addToSet': { category: ObjectId(email_id) } })

def findById(db, email_id):
    email = db.emails.find_one({ '_id': ObjectId(email_id) })
    return parse_mongo(email)

def addToDraft(db, user_id, email_id):
    db.users.update_one({ '_id': ObjectId(user_id) }, { '$addToSet': { 'draft': ObjectId(email_id) } })

def deleteDraft(db, user_id, email_id):
    db.users.update_one({ '_id': ObjectId(user_id) }, { '$pull': { 'draft': ObjectId(email_id) } })
    db.emails.remove({'_id': ObjectId(email_id)})

def sendEmail(db, email_id):
    email = findById(db, email_id)
    email = parse_mongo(email)
    if not email['to_user']:
        return
    db.emails.update_one({ '_id': ObjectId(email_id) }, { '$set': { 'is_draft': False }})
    to_user_id = email['to_user']['$oid']
    from_user_id = email['from_user']['$oid']

    db.users.update_one({ '_id': ObjectId(from_user_id) }, { '$pull': { 'draft': ObjectId(email_id) } })

    db.users.update_one({ '_id': ObjectId(from_user_id) }, { '$addToSet': { 'sent': ObjectId(email_id) } })
    db.users.update_one({ '_id': ObjectId(to_user_id) }, { '$addToSet': { 'inbox': ObjectId(email_id) } })

File: client/src/pages/Email.js
import React, { useContext, useEffect, useState } from 'react';
import { GlobalContext } from '../context';

import {Row, Col, message, Spin, Typography, Divider, Button, Dropdown, Menu, Form, Input} from 'antd';
import { DownOutlined } from '@ant-design/icons';

import {baseURI} from '../config';

import axios from 'axios';
import _ from 'lodash';

import {NavLink, Redirect} from 'react-router-dom';

const {Title} = Typography;

const categories = {
    'inbox': 'Inbox',
    'starred': 'Starred',
    'spam': 'Spam',
    'important': 'Important',
    'sent': 'Sent',
    'trash': 'Trash',
    'draft': 'Draft',
};

const EmailPage = (props) => {

    const { state, fetchProfile }  = useContext(GlobalContext);
    const [category, setCategory] = useState(null);
    const [folder, setFolder] = useState(null);
    const [email, setEmail] = useState(null);
    const [deleted, setDeleted] = useState(false);

    const checkCategoryAndFolder = (currentEmail) => {
        let found = false;
        _.forEach(categories, (value, key) => {
            const matched = _.find(state.user[key], mail => mail['$oid'] === currentEmail['_id']['$oid']);
            if (matched) {
                setCategory(key);                
                found = true;
            }
        });
        if (found) return;
        _.forEach(state.user.folders, currentFolder => {
            const matched = _.find(currentFolder.emails, mail => mail['$oid'] === currentEmail['_id']['$oid']);
            if (matched) {
                setFolder(currentFolder);
            }
        });
    }

    const fetchEmail = () => {
        const {params} = props.match;
        const {emailId} = params;
        const token = localStorage.getItem('jwt');
        axios.get(`${baseURI}/email/${emailId}`, {headers:{ 'Authorization': token }})
        .then((response) => {
            setEmail(response.data.email);
            checkCategoryAndFolder(response.data.email);
        }, (error) => {
            message.error(error.response.data.message);
        });
    }

    useEffect(() => {
        fetchEmail();
    }, []);

    const moveToCategory = (cat) => {
        const {params} = props.match;
        const {emailId} = params;
        const token = localStorage.getItem('jwt');
        axios.post(`${baseURI}/email/mark/${emailId}/${cat}`, {}, {headers:{ 'Authorization': token }})
        .then((response) => {
            message.success(response.data.message);
            fetchEmail();
            fetchProfile();
        }, (error) => {
            message.error(error.response.data.message);
        });
    }

    const moveToFolder = (fold) => {
        const {params} = props.match;
        const {emailId} = params;
        const folderId = fold['_id']['$oid'];
        const token = localStorage.getItem('jwt');
        axios.post(`${baseURI}/email/folder/${emailId}/${folderId}`, {}, {headers:{ 'Authorization': token }})
        .then((response) => {
            message.success(response.data.message);
            fetchEmail();
            fetchProfile();
        }, (error) => {
            message.error(error.response.data.message);
        });
    }

    const onDelete = () => {
        const {params} = props.match;
        const {emailId} = params;
        const token = localStorage.getItem('jwt');
        axios.delete(`${baseURI}/email/${emailId}`, {headers:{ 'Authorization': token }})
        .then((response) => {
            message.success(response.data.message);
            fetchProfile();
            setDeleted(true);
        }, (error) => {
            message.error(error.response.data.message);
        });
    }

    if (deleted)
        return <Redirect to="/" />;

    return (
        <> 
            {email && <div style={{margin: '20px'}}>
                <Row>
                    {_.map(categories, (value, key) => {
                        if (category !== key) {
                            return (
                                <Col span={3}>
                                    <Button type="primary" onClick={() => moveToCategory(key)}>Move to {categories[key]}</Button> 
                                </Col>
                            );
                        }
                        return null;
                    })}

                    <Col span={3}>
                        <Dropdown
                            overlay = {
                                <Menu>
                                    {state.folders.map(fold => {

                                        if (folder && folder.folder_name === fold.folder_name) return null;

                                        return (
                                            <Menu.Item>
                                                <Button onClick={() => moveToFolder(fold)} >{fold.folder_name}</Button>
                                            </Menu.Item>
                                        );
                                    })}
                                </Menu>
                            }
                        >
                            <a className="ant-dropdown-link">
                                Move To <DownOutlined />
                            </a>
                        </Dropdown>
                    </Col>
                </Row>
                <Divider />
                <Row offset={8} span={12}>
                    <Col span={4} offset={4}>
                        <Title level={5} >To:</Title>
                    </Col>
                    <Col >
                        <Title level={5} >{_.invert(state.users)[email['to_user']['$oid']]}</Title>
                    </Col>
                </Row>
                <Divider />
                <Row offset={8} span={12}>
                    <Col span={4} offset={4}>
                        <Title level={5} >From:</Title>
                    </Col>
                    <Col >
                        <Title level={5} >{_.invert(state.users)[email['from_user']['$oid']]}</Title>
                    </Col>
                </Row>
                <Divider />
                <Row offset={8} span={12}>
                    <Col span={4} offset={4}>
                        <Title level={5} >Subject:</Title>
                    </Col>
                    <Col >
                        <Title level={5} >{email.subject}</Title>
                    </Col>
                </Row>
                <Divider />
                <Row offset={8} span={12}>
                    <Col span={4} offset={4}>
                        <Title level={5} >Body:</Title>
                    </Col>
                    <Col >
                        <Title level={5}>{email.content}</Title>
                    </Col>
                </Row>
                <Divider />

                <Button type="primary" onClick={onDelete} >
                    Delete Email
                </Button>
            </div>}
            {!email && <div style={{marginTop: '35vh', textAlign: 'center'}}>
                <Spin />
            </div>}
        </>
    );
}

export default EmailPage;